server/src/main/java/io/crate/execution/engine/aggregation/impl/TDigestState.java

```diff
@@ -61,33 +59,29 @@ void fractions(double[] fractions) {
         this.fractions = fractions;
     }
 
-    int addGetSizeDelta(double value) {
-        add(value);
-        int newByteSize = byteSize();
-        int delta = newByteSize - lastByteSize;
-        lastByteSize = newByteSize;
-        return delta;
-    }
-
-
     public static void write(TDigestState state, StreamOutput out) throws IOException {
-        out.writeDouble(state.compression());
         out.writeDoubleArray(state.fractions);
-        out.writeVInt(state.centroidCount());
-        for (Centroid centroid : state.centroids()) {
-            out.writeDouble(centroid.mean());
-            out.writeVLong(centroid.count());
-        }
+
+        byte[] bytes = new byte[state.byteSize()];
+        ByteBuffer buf = ByteBuffer.wrap(bytes);
+        state.asBytes(buf);
+
+        out.writeVInt(bytes.length);
+        out.writeBytes(bytes, 0, bytes.length);
     }
 
     public static TDigestState read(StreamInput in) throws IOException {
-        double compression = in.readDouble();
-        double[] fractions = in.readDoubleArray();
-        TDigestState state = new TDigestState(compression, fractions);
-        int n = in.readVInt();
-        for (int i = 0; i < n; i++) {
-            state.add(in.readDouble(), in.readVInt());
-        }
-        return state;
+        double[] factions = in.readDoubleArray();
+
+        int len = in.readVInt();
+        byte[] bytes = new byte[len];
+        in.readBytes(bytes, 0, len);
+
+        // MergingDigest.fromBytes() restored digest is exact, but tDigestState.add() re-merges and may change centroid layout
+        MergingDigest mergingDigest = MergingDigest.fromBytes(ByteBuffer.wrap(bytes));
+        TDigestState tDigestState = new TDigestState(mergingDigest.compression(), factions);
+        tDigestState.add(mergingDigest);
```

Comment:
Another option is to change `TDigestState` to contain a `MergingDigest` as a delegate rather than extending it. Then the centroids can be restored exactly.